constraint/blueprint_scs.go

```diff
@@ -0,0 +1,299 @@
+package constraint
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/consensys/gnark/debug"
+)
+
+var (
+	errDivideByZero  = errors.New("division by 0")
+	errBoolConstrain = errors.New("boolean constraint doesn't hold")
+)
+
+// BlueprintGenericSparseR1C implements Blueprint and BlueprintSparseR1C.
+// Encodes
+//
+//	qL⋅xa + qR⋅xb + qO⋅xc + qM⋅(xaxb) + qC == 0
+type BlueprintGenericSparseR1C struct{}
+
+func (b *BlueprintGenericSparseR1C) NbInputs() int {
+	return 9 // number of fields in SparseR1C
+}
+func (b *BlueprintGenericSparseR1C) NbConstraints() int {
+	return 1
+}
+
+func (b *BlueprintGenericSparseR1C) CompressSparseR1C(c *SparseR1C) []uint32 {
+	return []uint32{
+		c.XA,
+		c.XB,
+		c.XC,
+		c.QL,
+		c.QR,
+		c.QO,
+		c.QM,
+		c.QC,
+		uint32(c.Commitment),
+	}
+}
+
+func (b *BlueprintGenericSparseR1C) DecompressSparseR1C(c *SparseR1C, calldata []uint32) {
+	c.Clear()
+
+	c.XA = calldata[0]
+	c.XB = calldata[1]
+	c.XC = calldata[2]
+	c.QL = calldata[3]
+	c.QR = calldata[4]
+	c.QO = calldata[5]
+	c.QM = calldata[6]
+	c.QC = calldata[7]
+	c.Commitment = CommitmentConstraint(calldata[8])
+}
+
+func (b *BlueprintGenericSparseR1C) Solve(s Solver, calldata []uint32) error {
+	var c SparseR1C
+	b.DecompressSparseR1C(&c, calldata)
+	if c.Commitment != NOT {
+		// a constraint of the form f_L - PI_2 = 0 or f_L = Comm.
+		// these are there for enforcing the correctness of the commitment and can be skipped in solving time
+		return nil
+	}
+
+	var ok bool
+
+	// constraint has at most one unsolved wire.
+	if !s.IsSolved(c.XA) {
+		// we solve for L: u1L+u2R+u3LR+u4O+k=0 => L(u1+u3R)+u2R+u4O+k = 0
+		if debug.Debug {
+			if !s.IsSolved(c.XB) || !s.IsSolved(c.XC) {
```

Comment:
I think compiler is smart enough since debug.Debug is a const set w/ a build tag, to omit completely that part. 

There is no particular risk even if there was a bug at this level; actually you made me realize this is useless, because Solver. GetValue already does this check. (even if it didn't and we somehow misordered solving, it wouldn't matter..)

Also, re-reading this part it seems there is a mistake 
```
		s.SetValue(c.XA, num)
``` 
when solving for `c.XB` (meaning we don't take that route often ^^)